### PR TITLE
Add biom tools: convert and add-metadata

### PR DESCRIPTION
**Pull request by blankenberg**

### CODE REVIEW

File: tools/biom_format/tool_dependencies.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+<tool_dependency>
+    <!-- <package name="python" version="2.7.10">
+        <repository name="package_python_2_7_10" owner="iuc" prior_installation_required="True" />
```
Comment (bgruening) [inline tools/biom_format/tool_dependencies.xml]: Is this WIP?

File: tools/biom_format/test-data/input_abundance_1.tabular [hunk 2]
```diff
@@ -0,0 +1,1834 @@
+2	0	0
```
Comment (bgruening) [inline tools/biom_format/test-data/input_abundance_1.tabular]: Would it make sense to only use `contains` as test method and keep this file smaller?

File: tools/biom_format/biom_convert.xml [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  --header-key TEXT               The observation metadata to include from the input BIOM table file when creating a tsv table file. By default no observation metadata will be included.
+  --output-metadata-id TEXT       The name to be given to the observation metadata column when creating a tsv table file if the column should be renamed.
+  --table-type                    OTU table,Pathway table,Function table,Ortholog table,Gene table,Metabolite table,Taxon table,Table The type of the table.
+  --process-obs-metadata          taxonomy,naive,sc_separated Process metadata associated with observations when converting from a classic table.
+  --tsv-metadata-formatter        naive,sc_separated Method for formatting the observation 
+  --help                          Show this message and exit.
+
+    ]]></help>
+    <tests>
+        <test>
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Test sections are coming before help. I think planemo lint will complain here.

File: tools/biom_format/biom_convert.xml [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    </inputs>
+    <outputs>
+        <data format="biom1" name="output_table">
+            <change_format>
+                <when input="output_type.output_type_selector" value="tsv" format="tabular" />
+                <when input="output_type.biom_type" value="hdf5" format="biom2" />
+            </change_format>
+        </data>
+    </outputs>
+    <tests>
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Oh here are the tests from below :)

File: tools/biom_format/biom_convert.xml [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                        <option value="hdf5">HDF5-formatted table</option>
+                    </param>
+                    <when value="json">
+                    </when>
+                    <when value="hdf5">
+                        <param name="collapsed_samples" type="boolean" truevalue="" falsevalue="--collapsed-samples" checked="False" label="Use collapsed samples"/>
+                        <param name="collapsed_observations" type="boolean" truevalue="" falsevalue="--collapsed-samples" checked="False" label="Use collapsed observations"/>
+                    </when>
+                </conditional>
+                    <param name="sample_metadata_fp" type="data" format="tabular" optional="True" label="Add sample metadata from metadata mapping file"/>
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Indentation.

File: tools/biom_format/biom_convert.xml [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                    <option value="Ortholog table">Ortholog table</option>
+                    <option value="Gene table">Gene table</option>
+                    <option value="Metabolite table">Metabolite table</option>
+                    <option value="Taxon table">Taxon table</option>
+                    <option value="Table" selected="True">Table</option>
+                </param>
+                <conditional name="biom_type">
+                    <param name="biom_type_selector" type="select" label="BIOM format type">
+                        <option value="json" selected="True">JSON-formatted table</option>
+                        <option value="hdf5">HDF5-formatted table</option>
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Maybe we can add here (bom2) and to the other one (biom1)

File: tools/biom_format/biom_convert.xml [hunk 7]
```diff
@@ -0,0 +1,169 @@
+<tool id="biom_convert" name="Convert BIOM" version="0.0.1">
+    <description>formats</description>
+    <requirements>
+        <requirement type="package" version="2.7.10">python</requirement>
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Should be enough to include biom-format.

File: tools/biom_format/biom_convert.xml [hunk 8]
```diff
@@ -0,0 +1,169 @@
+<tool id="biom_convert" name="Convert BIOM" version="0.0.1">
```
Comment (bgruening) [inline tools/biom_format/biom_convert.xml]: Should we start with version 2.1.5?

File: tools/biom_format/biom_add_metadata.xml [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  --help                          Show this message and exit.
+
+
+    ]]></help>
+    <tests>
+        <test>
+            <param name="input_table" value="input_abundance_1.biom1" ftype="biom1"/>
+            <param name="input_observation_metadata" value="input_taxonomy_1.tabular" ftype="tabular"/>
+            <param name="sc_separated" value="taxonomy"/>
+            <param name="input_sample_metadata" value=""/>
```
Comment (bgruening) [inline tools/biom_format/biom_add_metadata.xml]: Tests needs to move up.

File: tools/biom_format/biom_add_metadata.xml [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                #end if
+                #if $observation_header:
+                    --observation-header "${observation_header}"
+                #end if
+                ${output_as_json}                
+        ]]>
+    </command>
+    <inputs>
+        <param name="input_table" type="data" format="biom1" label="BIOM File"/>
+        <param name="input_sample_metadata" type="data" format="tabular" optional="True" label="Sample Metadata Tabular File"/>
```
Comment (bgruening) [inline tools/biom_format/biom_add_metadata.xml]: There is no help="" section that indicates the original parameter. Without this the help is kind of useless.
Maybe argument="--observation-metadata-fp" would be useful for al parameters?

File: tools/biom_format/biom_add_metadata.xml [hunk 11]
```diff
@@ -0,0 +1,137 @@
+<tool id="biom_add_metadata" name="BIOM metadata" version="0.0.1">
+    <description>add-metadata</description>
+    <requirements>
+        <requirement type="package" version="2.7.10">python</requirement>
```
Comment (bgruening) [inline tools/biom_format/biom_add_metadata.xml]: python requirement can be removed

**Comment by bgruening** [comment]
Tests are failing with `RunToolException: Error creating a job for these tool inputs - History does not include a dataset of the required format / build`.
We need to get the Biom format into Galaxy 16.04 ... mh. Any idea?

**Comment by bgruening** [comment]
@blankenberg thanks for working on this!

**Comment by blankenberg** [comment]
@bgruening thanks for the thorough review. Tests should pass once https://github.com/galaxyproject/galaxy/issues/2423 is fixed, and passed locally with 
```
$ git diff
diff --git a/lib/galaxy/web/security/__init__.py b/lib/galaxy/web/security/__init__.py
index 29726ee..5216fe7 100644
--- a/lib/galaxy/web/security/__init__.py
+++ b/lib/galaxy/web/security/__init__.py
@@ -44,8 +44,8 @@ class SecurityHelper( object ):
         self.id_ciphers_for_kind = _cipher_cache( per_kind_id_secret_base )
 
     def encode_id( self, obj_id, kind=None ):
-        if obj_id is None:
-            raise galaxy.exceptions.MalformedId("Attempted to encode None id")
+        #if obj_id is None:
+        #    raise galaxy.exceptions.MalformedId("Attempted to encode None id")
         id_cipher = self.__id_cipher( kind )
         # Convert to string
         s = str( obj_id )
```

File: tools/biom_format/tool_dependencies.xml [hunk 12]
```diff
@@ -1,6 +0,0 @@
-<?xml version="1.0"?>
```
Comment (bgruening) [inline tools/biom_format/tool_dependencies.xml]: I'm fine with traditional tool_dependency especially if they are already available :) Any reason to remove it?

**Comment by bgruening** [comment]
@blankenberg this looks fantastic. Let's see when the bug is fixed. Thanks!

**Comment by bgruening** [comment]
Restarted tests as your patch was applied.

**Comment by bgruening** [comment]
Tada! Thanks @blankenberg!

**Comment by blankenberg** [comment]
Thanks!

**Pull request merged**